org.eclipse.jdt.ui/core refactoring/org/eclipse/jdt/internal/corext/refactoring/util/SideEffectChecker.java

```diff
@@ -0,0 +1,324 @@
+ *******************************************************************************/
+package org.eclipse.jdt.internal.corext.refactoring.util;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+
+import org.eclipse.jdt.core.ICompilationUnit;
+import org.eclipse.jdt.core.IMethod;
+import org.eclipse.jdt.core.IType;
+import org.eclipse.jdt.core.ITypeHierarchy;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.core.dom.AST;
+import org.eclipse.jdt.core.dom.ASTMatcher;
+import org.eclipse.jdt.core.dom.ASTNode;
+import org.eclipse.jdt.core.dom.ASTParser;
+import org.eclipse.jdt.core.dom.ASTVisitor;
+import org.eclipse.jdt.core.dom.ArrayAccess;
+import org.eclipse.jdt.core.dom.Assignment;
+import org.eclipse.jdt.core.dom.CastExpression;
+import org.eclipse.jdt.core.dom.CompilationUnit;
+import org.eclipse.jdt.core.dom.Expression;
+import org.eclipse.jdt.core.dom.FieldAccess;
+import org.eclipse.jdt.core.dom.IMethodBinding;
+import org.eclipse.jdt.core.dom.ITypeBinding;
+import org.eclipse.jdt.core.dom.IVariableBinding;
+import org.eclipse.jdt.core.dom.MethodDeclaration;
+import org.eclipse.jdt.core.dom.MethodInvocation;
+import org.eclipse.jdt.core.dom.Modifier;
+import org.eclipse.jdt.core.dom.NodeFinder;
+import org.eclipse.jdt.core.dom.ParenthesizedExpression;
+import org.eclipse.jdt.core.dom.PostfixExpression;
+import org.eclipse.jdt.core.dom.PrefixExpression;
+import org.eclipse.jdt.core.dom.QualifiedName;
+import org.eclipse.jdt.core.dom.SimpleName;
+
+import org.eclipse.jdt.internal.corext.dom.ASTNodes;
+import org.eclipse.jdt.internal.corext.util.JavaModelUtil;
+
+public class SideEffectChecker extends ASTVisitor {
+
+	public SideEffectChecker(ASTNode astNode) {
+		fExpression= astNode;
+		fSideEffect= false;
+	}
+
+	boolean fSideEffect;
+
+	ASTNode fExpression;
+
+	private final int THRESHOld= 1000;
+
+	static private HashSet<String> MARKEDMETHODSET= new HashSet<>();
+	static {
+		String[] methodNames= { "java.lang.System.currentTimeMillis", "java.lang.System.nanoTime", //$NON-NLS-1$ //$NON-NLS-2$
+				"java.io.PrintStream.print", "java.io.PrintStream.printf", //$NON-NLS-1$ //$NON-NLS-2$
+				"java.io.PrintStream.println" }; //$NON-NLS-1$
+		MARKEDMETHODSET.addAll(Arrays.asList(methodNames));
+	}
+
+
+	public boolean hasSideEffect() {
+		return fSideEffect;
+	}
+
+	@Override
+	public boolean preVisit2(ASTNode node) {
+		if (hasSideEffect()) {
+			return false;
+		}
+		if (selfModied(node)) {
+			fSideEffect= true;
+			return false;
+		}
+		if (node instanceof MethodInvocation) {
+			MethodInvocation node2= (MethodInvocation) node;
+			final IMethodBinding resolveMethodBinding= node2.resolveMethodBinding();
+			if (MARKEDMETHODSET.contains(getQualifiedName(resolveMethodBinding))) {
+				fSideEffect= true;
+				return false;
+			}
+			MethodDeclaration md= findFunctionDefinition(resolveMethodBinding.getDeclaringClass(), resolveMethodBinding);
```

Comment:
I think if you cannot parse a function and it isn't part of a set you have pre-cleared  you should assume side-effects and only change one instance.  This is safe and you are at least changing one instance.

Reply:
Ok now.